### DISCUSSION TITLE
LocalChannel for WebApi

### DISCUSSION THREAD

**Discussion by rhyous** [dotnet/aspnetcore#23918] [General]
I would like to call a webapi service, webapi2, from another webapi service, webapi1. That sounds easy easy enough. However, the Filters and pipeline don't get called when calling webapi2 from webapi1. However, for my application's architecture, I would like to call webap 2 _and_ get the Filters and pipeline to apply on inter-service calls.

HttpClient works, but the overhead is pretty high. Too high, in fact. It results in serialization and deserialization, leaving to the OS' network, creating sockets, etc. So I want a local call that bypasses serialization, and stays in the code, much like how WCF had a local channel. https://docs.microsoft.com/en-us/dotnet/framework/wcf/samples/local-channel

If the code doesn't exist, but extensibility is there, I'm willing to write it, if someone can point me to the right interfaces, etc.

**Comment by Tratcher** [comment]
That doesn't exist as a pre-built feature, but most of what you'd need already exists as part of the TestServer. It even has an HttpClient API that bypasses sockets to communicate in-memory.

Here are the raw components I expect you could copy and reassemble:
https://github.com/dotnet/aspnetcore/blob/f0bb1315cc057a6b7f17418c48762d51c0218a0f/src/Hosting/TestHost/src/TestServer.cs#L102-L160
https://github.com/dotnet/aspnetcore/blob/f0bb1315cc057a6b7f17418c48762d51c0218a0f/src/Hosting/TestHost/src/ClientHandler.cs
https://github.com/dotnet/aspnetcore/blob/f0bb1315cc057a6b7f17418c48762d51c0218a0f/src/Hosting/TestHost/src/HttpContextBuilder.cs

I'd expect you to convert TestServer from an IServer to a middleware so it could inject requests into the pipeline. You'll also probably need a singleton DI service for the TestServer to provide access to its HttpClient.

I'm curious to see the results.

**Reply by rhyous** [reply]
Thanks. I'll give it a look.
